cli_tools/google-osconfig-agent/config/config.go

```diff
@@ -37,19 +51,119 @@ var (
-	yumRepoFilePath    = flag.String("yum_repo_file", "/etc/yum/repos.d/google_osconfig.repo", "yum repo file location")
-	aptRepoFilePath    = flag.String("apt_repo_file", "/etc/apt/sources.list.d/google_osconfig.list", "apt repo file location")
+	agentConfig   *config
+	agentConfigMx sync.RWMutex
+	version       string
+)
 
-	osPackageEnabled bool
-	osPatchEnabled   bool
+type config struct {
+	osInventoryEnabled, osPackageEnabled, osPatchEnabled                     bool
+	googetRepoFilePath, zypperRepoFilePath, yumRepoFilePath, aptRepoFilePath string
+}
 
-	version string
-)
+func getAgentConfig() config {
+	agentConfigMx.RLock()
+	defer agentConfigMx.RUnlock()
+	return *agentConfig
+}
+
+func enabled(s string) bool {
+	enabled, err := strconv.ParseBool(s)
+	if err != nil {
+		// Bad entry returns as not enabled.
+		return false
+	}
+	return enabled
+}
+
+type metadataJSON struct {
+	Instance instanceJSON
+	Project  projectJSON
+}
+
+type instanceJSON struct {
+	Attributes attributesJSON
+}
+
+type projectJSON struct {
+	Attributes attributesJSON
+}
+
+type attributesJSON struct {
+	OSInventoryEnabled string `json:"os-inventory-enabled"`
+	OSPatchEnabled     string `json:"os-patch-enabled"`
+	OSPackageEnabled   string `json:"os-package-enabled"`
+}
+
+func (c *config) setFromMetadata(md metadataJSON) {
+	// Check project first then instance as instance metadata overrides project.
+	if md.Project.Attributes.OSInventoryEnabled != "" {
+		c.osInventoryEnabled = enabled(md.Project.Attributes.OSInventoryEnabled)
+	}
+	if md.Project.Attributes.OSPatchEnabled != "" {
+		c.osPatchEnabled = enabled(md.Project.Attributes.OSPatchEnabled)
+	}
+	if md.Project.Attributes.OSPackageEnabled != "" {
+		c.osPackageEnabled = enabled(md.Project.Attributes.OSPackageEnabled)
+	}
+
+	if md.Instance.Attributes.OSInventoryEnabled != "" {
+		c.osInventoryEnabled = enabled(md.Instance.Attributes.OSInventoryEnabled)
+	}
+	if md.Instance.Attributes.OSPatchEnabled != "" {
+		c.osPatchEnabled = enabled(md.Instance.Attributes.OSPatchEnabled)
+	}
+	if md.Instance.Attributes.OSPackageEnabled != "" {
+		c.osPackageEnabled = enabled(md.Instance.Attributes.OSPackageEnabled)
+	}
+}
+
+// SetConfig sets the agent config.
+func SetConfig() error {
+	agentConfigMx.Lock()
+	defer agentConfigMx.Unlock()
+	agentConfig = &config{
+		osInventoryEnabled: osInventoryEnabled,
+		osPackageEnabled:   osPackageEnabled,
+		osPatchEnabled:     osPatchEnabled,
+		googetRepoFilePath: googetRepoFilePath,
+		zypperRepoFilePath: zypperRepoFilePath,
+		yumRepoFilePath:    yumRepoFilePath,
+		aptRepoFilePath:    aptRepoFilePath,
+	}
+
+	var md string
+	var err error
+	webError := 0
+	for {
+		md, err = metadata.Get("?recursive=true&alt=json")
+		if err == nil {
+			break
+		}
+		// Only log the third web error to avoid transient errors and
+		// not to spam the log on network failures.
```

Comment:
This comment doesn't match what the code is doing and I'm not sure if the comment is out of date or if it points to a bug. First, the special logic happens on the 2nd web error and not the 3rd. But also, this function doesn't log but exits.

So, should it handle the error on the 2nd or 3rd error? And should the comment be changed to "Only return the third web error..."?

Also, did you chose 2nd (or 3rd) error because usually metadata is only down for <= 15 seconds?

Reply:
Fixed, this was not doing what I wanted